rust/connlib/libs/tunnel/src/tun_android.rs

```diff
@@ -28,30 +29,33 @@ impl Drop for IfaceDevice {
 }
 
 impl IfaceDevice {
-    fn write(&self, _buf: &[u8]) -> usize {
-        tracing::error!("`write` unimplemented on Android");
-        0
+    fn write(&self, buf: &[u8]) -> usize {
+        match unsafe { write(self.fd, buf.as_ptr() as _, buf.len() as _) } {
+            -1 => 0,
+            n => n as usize,
+        }
     }
 
-    pub async fn new() -> Result<Self> {
-        // TODO: This won't actually work for non-root users...
-        let fd = unsafe { open(b"/dev/net/tun\0".as_ptr() as _, O_RDWR) };
-        // TODO: everything!
-        if fd == -1 {
-            Err(Error::Io(std::io::Error::last_os_error()))
-        } else {
-            Ok(Self { fd })
-        }
+    pub async fn new(fd: Option<i32>) -> Result<Self> {
+        log::debug!("tunnel allocation unimplemented on Android; using provided fd");
+        Ok(Self {
+            fd: fd.expect("file descriptor must be provided!") as RawFd,
+        })
     }
 
     pub fn set_non_blocking(self) -> Result<Self> {
-        tracing::error!("`set_non_blocking` unimplemented on Android");
+        // Anrdoid already opens the tun device in non-blocking mode for us
+        log::debug!("`set_non_blocking` unimplemented on Android");
         Ok(self)
     }
 
     pub async fn mtu(&self) -> Result<usize> {
-        tracing::error!("`mtu` unimplemented on Android");
-        Ok(0)
+        // We stick with a hardcoded MTU of 1280 for now. This could be improved by
+        // finding the MTU of the underlying physical interface and subtracting 80
+        // from it for the WireGuard overhead, but that's a lot of complexity
+        // for little gain.
```

Comment:
we still should find the actual mtu or at least have the callbacks tell us the mtu that it used at some point, since if it configured a great MTU we could read packets partially.

Reply:
also note that this should be the wireguard's interface mtu not the physical interface, each peer could use a different physical interface and it can also change.

Reply:
It's a little involved in Android since we first have to retrieve the interface name with `TUNGETIFF`. But I'll take that as an exercise for the reader.

Reply:
I'll save this for the next PR if that's alright